### ISSUE TITLE
Upgrade \select@language to a real hook

### ISSUE THREAD

**Issue by moewew** [reutenauer/polyglossia#398]
See https://github.com/plk/biblatex/issues/970 and https://github.com/latex3/babel/issues/55 for context.

If I understand correctly, currently `\select@language` is only kept around for `biblatex` compatibility

https://github.com/reutenauer/polyglossia/blob/7bcf51db90a4baa9f3cffa237f2ef518753e7c17/tex/polyglossia.sty#L2022-L2046

Would it be possible to 'upgrade'  `\select@language` to an official interface, maybe embedded into a scheme similar to `babel`'s `babel` hooks?

MWE to play around
```latex
\documentclass{article}
\usepackage[T1]{fontenc}
\usepackage[utf8]{inputenc}

\usepackage{etoolbox}

\usepackage{polyglossia}
\setdefaultlanguage[variant=british]{english}
\setotherlanguage{dutch}
\setotherlanguage[spelling=new]{german}
\setotherlanguage{hungarian}
\setotherlanguage{french}

\usepackage{xcolor}

\makeatletter
\pretocmd\select@language{\myfakehook{#1}}
  {}{\fail}

\newcommand*{\myfakehook}[1]{%
  \edef\mylanguagename{#1}%
  \begingroup
  \edef\my@tempa{%
    \endgroup
    \csgappto{blockextras@bbl@#1}{%
      \noexpand\typeout{selecting #1}%
      \noexpand\@ifundefinedcolor{col@#1}
        {\noexpand\normalcolor}
        {\noexpand\color{col@#1}}}%
    \csgappto{inlinextras@bbl@#1}{%
      \noexpand\typeout{selecting #1}%
      \noexpand\@ifundefinedcolor{col@#1}
        {\noexpand\normalcolor}
        {\noexpand\color{col@#1}}}%
    \csgappto{noextras@bbl@#1}{%
      \noexpand\typeout{deselecting #1}%
    }}%
  \my@tempa
}

\definecolor{col@british}{rgb}{0,0,1}
\definecolor{col@ngerman}{rgb}{1,0,1}
\definecolor{col@dutch}{HTML}{FFA500}
\makeatother

\begin{document}
Lorem \mylanguagename\ (british)

\begin{german}
ipsum \mylanguagename\ (ngerman)
\end{german}

\begin{french}
dolor \mylanguagename\ (french)
\end{french}

\begin{dutch}
sit \mylanguagename\ (dutch)
\end{dutch}

% the uses of \foreignlanguage are interesting:
% they call \extras<...>
% but not \select@language
% so the outcome depends on whether or not
% the language has been used (with \select@language) before

% was used before: extras contain something
% and are used, but \mylanguagename is off
\textfrench{amet \mylanguagename\ (french)}

% wasn't used before: extras empty (so we keep)
\texthungarian{consectur \mylanguagename\ (magyar)}
\end{document}
```

I'm copying over the desiderata for the hook from https://github.com/latex3/babel/issues/55
* executed every time `polyglossia` switches (or selects) languages
* the hook knows the selected language (this can work via `\babelname`, so we only need to make sure the hook is executed when `\babelname` is already set to the correct value)
* the hook can write to `\blockextras@bbl@<babelname>`, `\inlineextras@bbl@<babelname>` and `\noextras@bbl@<babelname>`
* the hook is executed before  `\blockextras@bbl@<babelname>` or `\inlineextras@bbl@<babelname>`  is used
* ideally the hook is executed at a grouping level more or less equal to `\select@language` which means, we don't need `\global` to (re)define a macro for the duration of the selected language.

**Comment by jspitz** [comment]
I fear I do not completely follow you. Could you summarize a a few words what this requested hook should actually do?

**Comment by moewew** [comment]
I have tightened up the list of desiderata above a bit (I had originally copied it from my `babel` post; `\select@language` already satisfies most criteria listed there anyway, it would only have to be executed in `\xpg@otherlanguage` as well).

Essentially what I'm looking for is something that works like `\select@language` but has an official interface from the `polyglossia` side (preferably documented) that avoids us (`biblatex`) having to deploy `\pretocmd`, which is category-code sensitive.

At the moment I can see this going two ways
* The approach may look similar to `babel`'s `\AddBabelHook`, where we can define named hooks that are executed at certain trigger events.
* `\select@language` (or its hook replacement) becomes a real hook in the `etoolbox` sense, which means it becomes a simple macro without argument and can be filled with `\preto` instead of `\pretocmd`.

**Comment by jspitz** [comment]
> I have tightened up the list of desiderata above a bit (I had originally copied it from my `babel` post; `\select@language` already satisfies most criteria listed there anyway, it would only have to be executed in `\xpg@otherlanguage` as well).

We have the hook `\inlineextras@bbl@<babelname>` which is executed there

For the block commands, there is `\blockextras@bbl@<babelname>`

For noextras, we have `\noextras@bbl@<babelname>`

All these hooks are "official" and documented, see polyglossia manual sec 7.

Are these executed too late?

**Comment by moewew** [comment]
Yes, that's too late because I actually want to append something to those hooks. I'll edit the list of desiderata again to switch from the `babel` names `\extras<languagename>` to the `polyglossia` names.

One important thing for the hook I'm looking for is that I don't know the current language in advance.

**Comment by jspitz** [comment]
> Yes, that's too late because I actually want to append something to those hooks. 

What do you mean: you want too "append something to those hooks"? Do you mean "prepend"? And to which hooks exactly? If the one I mentioned, why not use (e.g.)

`\gpreto\inlineextras@bbl@<babelname>`

> One important thing for the hook I'm looking for is that I don't know the current language in advance.

But you know the language (babelname) for which the hook shall be executed?

Sorry for being so slow.

**Comment by jspitz** [comment]
Maybe to make things more clear: `\inlineextras@bbl@\babelname` is an extra hook (empty by default) that is executed always _after_ `\inlineextras@<language>` in `\xpg@otherlanguage`. The same applies to the `blockextras` and `noextras` hooks.

**Comment by moewew** [comment]
Append or prepend doesn't really matter. What matters is the following: In `biblatex` the language-specific code for each language has to be added to `\inlineextras@bbl@<babelname>` (and friends, but let's focus on that one for now). In order to do that `biblatex` needs to load its own language definitions. Of course that means that `biblatex` needs to know which language is being used. That's where the new hook (or currently `\select@language`) comes in: The code there tells `biblatex` which language is being loaded so it can populate `\inlineextras@bbl@<babelname>` accordingly, it also sets `biblatex`'s `\currentlang` (pretty much only a unified interface for `babel`'s `\languagename` and `polyglossia`'s `\babelname`).

Essentially `\select@language` says to `biblatex`: The language `<language>` was loaded, and then `biblatex` can go on loading its definitions for that language and add them to  `\inlineextras@bbl@<babelname>`.

**Comment by jspitz** [comment]
Would something like this work for you?

```diff
diff --git a/tex/polyglossia.sty b/tex/polyglossia.sty
index 1e872ab..f719663 100644
--- a/tex/polyglossia.sty
+++ b/tex/polyglossia.sty
@@ -1551,6 +1551,8 @@
   \polyglossia@error@iflangnotloaded:n{#2}
   \exp_args:Nne \setkeys{#2}{#1}%
   \polyglossia@setforeignlanguage[#1]{#2}
+  % Hook for biblatex
+  \xpg@bbl@selected{\babelname}%
   % buggy restoration heure
   \csuse{inlineextras@#2}%
   % This is a hook for external packages which want to access variants
@@ -2028,6 +2030,8 @@
    \polyglossia@lang@setpardirection:n{#2}%
    \csuse{captions#2}%
    \csuse{date#2}%
+   % Hook for biblatex
+   \xpg@bbl@selected{\babelname}%
    % These are hooks for external packages which want to access variants
    % via babelname (such as biblatex)
    \cs_if_exist_use:c{captions@bbl@\babelname}%
@@ -2041,8 +2045,9 @@
    \cs_if_exist_use:c{blockextras@bbl@\babelname}%
  }
 
-% hook for compatibility with biblatex
+% hooks for compatibility with biblatex
 \def\select@language#1{}
+\def\xpg@bbl@selected#1{}
 
 \def\noextrascurrent#1{%
    \csuse{noextras@#1}%
```

**Comment by moewew** [comment]
Almost. That is pretty similar to what we have already.

The disadvantage of the approach with an argument is that it is harder to append to a macro with arguments than to a macro without argument. If `\babelname` is available anyway, we don't win much by passing it on. So maybe something like this
```diff
--- polyglossia-orig.sty	2020-04-03 16:45:49.702070600 +0200
+++ polyglossia.sty	2020-04-03 16:47:50.015290500 +0200
@@ -1551,6 +1551,8 @@
   \polyglossia@error@iflangnotloaded:n{#2}
   \exp_args:Nne \setkeys{#2}{#1}%
   \polyglossia@setforeignlanguage[#1]{#2}
+  % Hook for biblatex
+  \xpg@bbl@selected
   % buggy restoration heure
   \csuse{inlineextras@#2}%
   % This is a hook for external packages which want to access variants
@@ -2025,6 +2027,8 @@
    \xpg@set@bbl@hyphendata{\the\language}
    \xpg@initial@setup%
    \select@@language[#1]{#2}%
+   % Hook for biblatex
+   \xpg@bbl@selected
    \polyglossia@lang@setpardirection:n{#2}%
    \csuse{captions#2}%
    \csuse{date#2}%
@@ -2043,6 +2047,7 @@
 
 % hook for compatibility with biblatex
 \def\select@language#1{}
+\def\xpg@bbl@selected{}
 
 \def\noextrascurrent#1{%
    \csuse{noextras@#1}%

```
Note that I'm **not** sure about the exact placement of the hook inside the macros. Ideally it would be as early as possible, but it must have `\babelname` available with the the value of the newly selected language. `\select@@language` seems to do some typesetting setup already, so maybe even earlier than that?


This would be the second option in what I earlier commented

> At the moment I can see this going two ways
>* The approach may look similar to `babel`'s `\AddBabelHook`, where we can define named hooks that are executed at certain trigger events.
>* `\select@language` (or its hook replacement) becomes a real hook in the `etoolbox` sense, which means it becomes a simple macro without argument and can be filled with `\preto` instead of `\pretocmd`.

The first approach is still very interesting, but would be more work.

**Comment by jspitz** [comment]
OK, please try from https://github.com/jspitz/polyglossia/commit/b6b8a452447c50cddda6c41fed2f1d987d465a64

The hook is as early as possible, as `\babelname` is set in `\select@@language` and `\polyglossia@setforeignlanguage`, respectively.

**Comment by moewew** [comment]
Works really well. Thank you very much.

Just a minor quibble about the comments and the commit message: My hope was that this could be a universal hook for all packages to use and not just a compatibility kludge like `\select@language`, so maybe it is wise not to focus on `biblatex` too much.

**Comment by jspitz** [comment]
These are just internal comments for us to know why we introduced it in the first place.

I will push some documentation which will document it as a universal hook for packages (this is already in my pipe, just not yet pushed).

**Comment by jspitz** [comment]
It's in with all documentation.

**Comment by moewew** [comment]
Thanks. `biblatex` is going to use it from now on https://github.com/plk/biblatex/commit/b159f0ea158afa38a6e679f324355abcb03fd7fd.

**Comment by jspitz** [comment]
Nice!

**Comment by reutenauer** [comment]
Thanks to you both!  Would you like me to make release?

**Comment by jspitz** [comment]
If `biblatex` release is not due soon, I'd wait a few days until TL 2020 is out to check and remove some (all?) LuaTeX warnings that have become obsolete with `hbluatex` (#337).

**Comment by moewew** [comment]
From the `biblatex` point of view there is no need to rush, there is still quite a few things to be done until the next `biblatex` release.
